In [2]:
import pandas as pd
import json

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Ajustar la opción para mostrar más caracteres en cada columna
pd.set_option('display.max_colwidth', None)

In [ ]:
# Abrir el archivo JSON y leer línea por línea
with open(r'C:\\Users\\Pablo\\OneDrive\\Henry\\ProyectoFinal\\Google Maps\\metadata-sitios\\1.json') as file:
    data = [json.loads(line) for line in file]

# Convertir la lista de diccionarios en un DataFrame
df1 = pd.DataFrame(data)

In [ ]:
# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\2.json'
# Leer el JSON en un DataFrame
df2 = pd.read_json(ruta_json, lines=True)

# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\3.json'
# Leer el JSON en un DataFrame
df3 = pd.read_json(ruta_json, lines=True)

# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\4.json'
# Leer el JSON en un DataFrame
df4 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\5.json'
# Leer el JSON en un DataFrame
df5 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\6.json'
# Leer el JSON en un DataFrame
df6 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\7.json'
# Leer el JSON en un DataFrame
df7 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\8.json'
# Leer el JSON en un DataFrame
df8 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\9.json'
# Leer el JSON en un DataFrame
df9 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\10.json'
# Leer el JSON en un DataFrame
df10 = pd.read_json(ruta_json, lines=True)

In [ ]:
def extractCity(df):
    """
    Esta función toma un DataFrame que contiene una columna llamada 'address',
    la cual tiene direcciones formateadas con comas. La función extrae el valor
    que está entre la segunda y tercera coma (que representa la ciudad) y lo
    almacena en una nueva columna llamada 'city'.

    Parámetros:
    df (DataFrame): El DataFrame que contiene la columna 'address'.

    Retorna:
    DataFrame: El DataFrame original con una nueva columna 'city' que contiene
               el valor entre la segunda y tercera coma de 'address'.
    """

    # Dividir la columna 'address' en partes usando la coma como delimitador
    # Luego, extraer el segundo elemento (índice 2), que está entre la segunda y tercera coma

    df['city'] = df['address'].str.split(',').str[2].str.strip()

    return df

In [ ]:
def nulosYVacios(df):
    """
    Este procedimiento recibe un DataFrame y muestra la cantidad de valores nulos, vacíos y la cantidad de
    valores duplicados en todas las columnas excepto en las 6, 10, 11 y 13 del DataFrame. También muestra
    el mínimo, máximo y media para columnas de tipo float e integer.

    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    """
    print("Informe de Nulos, Vacíos y Duplicados:")
    print("-" * 30)

    # Total de registros en el DataFrame
    total_registros = len(df)

    # Imprimir total de registros
    print(f"Total de registros: {total_registros}")
    print("-" * 30)

    # Iterar sobre todas las columnas para mostrar nulos y vacíos
    for columna in df.columns:
        # Contar valores nulos
        nulos = df[columna].isnull().sum()

        # Contar valores vacíos (considerando que solo se cuentan si son cadenas)
        vacios = (df[columna] == '').sum() if df[columna].dtype == 'object' else 0

        # Calcular porcentajes
        porcentaje_nulos = (nulos / total_registros) * 100 if total_registros > 0 else 0
        porcentaje_vacios = (vacios / total_registros) * 100 if total_registros > 0 else 0

        # Imprimir resultados
        print(f"Columna: {columna}")
        print(f"  Valores Nulos: {nulos}")
        print(f"  Valores Vacíos: {vacios}")
        print(f"  Porcentaje Nulos: {porcentaje_nulos:.2f}%")
        print(f"  Porcentaje Vacíos: {porcentaje_vacios:.2f}%")

        # Contar y mostrar la cantidad de valores duplicados solo si no es una de las columnas con datos anidados
        if columna not in [df.columns[6], df.columns[10], df.columns[11], df.columns[13]]:
            # Filtrar los valores no nulos antes de contar duplicados
            cantidad_duplicados = df[columna].dropna().duplicated(keep=False).sum()
            print(f"  Cantidad de valores duplicados (solo no nulos): {cantidad_duplicados}")

        # Mostrar mínimo, máximo y media si la columna es de tipo float o int
        if df[columna].dtype in ['float64', 'int64']:
            minimo = df[columna].min()
            maximo = df[columna].max()
            media = df[columna].mean()
            print(f"  Mínimo: {minimo}")
            print(f"  Máximo: {maximo}")
            print(f"  Media: {media:.2f}")

        print("-" * 30)


Análisis preliminar de calidad de datos

In [ ]:
# Obtener la ciudad donde se encuentra el sitio con reseña
extractCity(df1)
extractCity(df2)
extractCity(df3)
extractCity(df4)
extractCity(df5)
extractCity(df6)
extractCity(df7)
extractCity(df8)
extractCity(df9)
extractCity(df10)

In [ ]:
def rankingFrecuenciaCiudad(df, x=10):
    """
    Este procedimiento recibe un DataFrame y un número entero x.
    Muestra en pantalla el ranking de los primeros x elementos con mayor frecuencia en la columna 'city'.

    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    x (int): El número de elementos a mostrar en el ranking.
    """
    print(f"Ranking de los primeros {x} elementos más frecuentes en la columna 'city':")
    print("-" * 60)

    # Verifica si la columna 'city' existe en el DataFrame
    if 'city' in df.columns:
        # Obtener la frecuencia de cada valor en la columna 'city'
        frecuencias = df['city'].value_counts()

        # Obtener los primeros x elementos con mayor frecuencia
        ranking = frecuencias.head(x)

        # Mostrar el ranking
        for index, value in ranking.items():
            print(f"{index}: {value}")
    else:
        print("La columna 'city' no existe en el DataFrame.")

    print("-" * 60)

In [ ]:
def mostrarCantidadFilasDuplicadas(df):
    """
    Filtra las filas duplicadas en un DataFrame según columnas específicas,
    y muestra la cantidad de filas duplicadas en la salida de pantalla.

    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    """
    # Filtrar duplicados considerando las columnas específicas
    filas_duplicadas = df[df.duplicated(subset=[df.columns[i] for i in [0, 1, 2, 4, 5]], keep=False)]

    # Contar la cantidad de filas duplicadas
    cantidad_duplicados = len(filas_duplicadas)

    # Mostrar la cantidad de filas duplicadas
    print(f"Cantidad de filas duplicadas: {cantidad_duplicados}")

Frecuencia de recomendaciones por ciudad segun google Maps

In [ ]:
# Obtener la frecuencia de cada valor en la columna 'city'
rankingFrecuenciaCiudad(df1)

Ranking de los primeros 10 elementos más frecuentes en la columna 'city':
------------------------------------------------------------
Houston: 2599
New York: 1903
Chicago: 1799
Los Angeles: 1663
Brooklyn: 1550
San Antonio: 1200
Dallas: 1154
Miami: 1139
Las Vegas: 1089
Philadelphia: 1044
------------------------------------------------------------


In [ ]:
rankingFrecuenciaCiudad(df2)

In [ ]:
rankingFrecuenciaCiudad(df3)

In [ ]:
rankingFrecuenciaCiudad(df4)
rankingFrecuenciaCiudad(df5)
rankingFrecuenciaCiudad(df6)
rankingFrecuenciaCiudad(df7)
rankingFrecuenciaCiudad(df8)
rankingFrecuenciaCiudad(df9)
rankingFrecuenciaCiudad(df10)

Ranking de los primeros 10 elementos más frecuentes en la columna 'city':
------------------------------------------------------------
Houston: 2511
New York: 1855
Los Angeles: 1713
Chicago: 1679
Brooklyn: 1596
Miami: 1337
San Antonio: 1249
Las Vegas: 1229
Philadelphia: 1204
Dallas: 1143
------------------------------------------------------------
Ranking de los primeros 10 elementos más frecuentes en la columna 'city':
------------------------------------------------------------
Houston: 2504
New York: 1825
Chicago: 1763
Los Angeles: 1625
Brooklyn: 1610
San Antonio: 1299
Dallas: 1185
Miami: 1148
Phoenix: 1071
Philadelphia: 1051
------------------------------------------------------------
Ranking de los primeros 10 elementos más frecuentes en la columna 'city':
------------------------------------------------------------
Houston: 2598
Chicago: 1812
New York: 1810
Brooklyn: 1636
Los Angeles: 1401
San Antonio: 1366
Dallas: 1288
Miami: 1232
Las Vegas: 1172
Phoenix: 1119
------------------

In [ ]:
nulosYVacios(df1)

Informe de Nulos, Vacíos y Duplicados:
------------------------------
Total de registros: 275001
------------------------------
Columna: name
  Valores Nulos: 7
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 84218
------------------------------
Columna: address
  Valores Nulos: 10062
  Valores Vacíos: 0
  Porcentaje Nulos: 3.66%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 51265
------------------------------
Columna: gmap_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 53146
------------------------------
Columna: description
  Valores Nulos: 261846
  Valores Vacíos: 0
  Porcentaje Nulos: 95.22%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 10123
------------------------------
Columna: latitude
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 275001 entries, 0 to 275000
Data columns (total 16 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   name              274994 non-null  object 
 1   address           264939 non-null  object 
 2   gmap_id           275001 non-null  object 
 3   description       13155 non-null   object 
 4   latitude          275001 non-null  float64
 5   longitude         275001 non-null  float64
 6   category          272740 non-null  object 
 7   avg_rating        275001 non-null  float64
 8   num_of_reviews    275001 non-null  int64  
 9   price             13450 non-null   object 
 10  hours             192448 non-null  object 
 11  MISC              194972 non-null  object 
 12  state             195523 non-null  object 
 13  relative_results  238771 non-null  object 
 14  url               275001 non-null  object 
 15  city              261516 non-null  object 
dtypes: float64(3), int64

Puede ser que se encuentren valores repetidos de lugar en el mapa, porque quizas es un edificio, o antes estuvo otro, por lo tanto habra que ver si es que tienen el mismo nombre. coincidencia entre esas coolumnas.
 0   name              274994 non-null  object
 1   address           264939 non-null  object
 2   gmap_id           275001 non-null  object
 4   latitude          275001 non-null  float64
 5   longitude         275001 non-null  float64
 14  url               275001 non-null  object

In [ ]:
# Filas duplicas considerando las columnas 0 a 5
mostrarCantidadFilasDuplicadas(df1)

Cantidad de filas duplicadas: 53146


In [ ]:
# Cargar el archivo de business
dfYeplp = pd.read_pickle(r'C:\Users\Pablo\OneDrive\Henry\Proyecto Final\Yelp\business.pkl')

In [ ]:
dfYeplp.info()

<class 'pandas.core.frame.DataFrame'>
Index: 150346 entries, 0 to 150345
Data columns (total 28 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   business_id   150346 non-null  object
 1   name          150346 non-null  object
 2   address       150346 non-null  object
 3   city          150346 non-null  object
 4   state         150343 non-null  object
 5   postal_code   150346 non-null  object
 6   latitude      150346 non-null  object
 7   longitude     150346 non-null  object
 8   stars         150346 non-null  object
 9   review_count  150346 non-null  object
 10  is_open       150346 non-null  object
 11  attributes    136602 non-null  object
 12  categories    150243 non-null  object
 13  hours         127123 non-null  object
 14  business_id   5 non-null       object
 15  name          5 non-null       object
 16  address       5 non-null       object
 17  city          5 non-null       object
 18  state         5 non-null     

In [ ]:
# Filtrar las filas donde gmap_id está duplicado
filas_duplicadas = df1[df1['gmap_id'].duplicated(keep=False)]


In [ ]:
# Ordenar el DataFrame por la columna gmap_id
df_ordenado = filas_duplicadas.sort_values(by='gmap_id')


df_ordenado


In [ ]:
df1.iloc[:,[0,1,2,3,14]]

name  \
0                       Porter Pharmacy   
1                          City Textile   
2                          San Soo Dang   
3                          Nova Fabrics   
4                      Nobel Textile Co   
...                                 ...   
274996                     Maven Realty   
274997               Stone Castle Fence   
274998  Health Insurance Solutions Team   
274999          Bernard Building Center   
275000    HMR Motor Works and Auto Care   

                                                                       address  \
0                          Porter Pharmacy, 129 N Second St, Cochran, GA 31014   
1                        City Textile, 3001 E Pico Blvd, Los Angeles, CA 90023   
2                       San Soo Dang, 761 S Vermont Ave, Los Angeles, CA 90005   
3                          Nova Fabrics, 2200 E 11th St, Los Angeles, CA 90021   
4                        Nobel Textile Co, 719 E 9th St, Los Angeles, CA 90021   
...                                                                        ...   
274996      Maven Realty, 2180 McMillan Ave #71672, North Charleston, SC 29405   
274997     Stone Castle Fence, 124 Cypress Gardens Rd, Moncks Corner, SC 29461   
274998  Health Insurance Solutions Team, 6650 Rivers Ave, Charleston, SC 29406   
274999            Bernard Building Center, 27300 M-32, M-32, Hillman, MI 49746   
275000                            HMR Motor Works and Auto Care, Mio, MI 48647   

                                      gmap_id description  \
0       0x88f16e41928ff687:0x883dad4fd048e8f8        None   
1       0x80c2c98c0e3c16fd:0x29ec8a728764fdf9        None   
2       0x80c2c778e3b73d33:0xbdc58662a4a97d49        None   
3        0x80c2c89923b27a41:0x32041559418d447        None   
4       0x80c2c632f933b073:0xc31785961fe826a6        None   
...                                       ...         ...   
274996  0x88fe64e7daceaaab:0x101c046dcadcc9c3        None   
274997  0x88fe59f4602c95fd:0x57e2b1904d9e6949        None   
274998  0x88fe71f9617c15e1:0x89518c1d251706bf        None   
274999  0x4d34c7517d6e6e29:0x7fe553ee2f090cf0        None   
275000  0x88208f4360c053ab:0xe3d6883f19601a01        None   

                                                                                                                     url  
0       https://www.google.com/maps/place//data=!4m2!3m1!1s0x88f16e41928ff687:0x883dad4fd048e8f8?authuser=-1&hl=en&gl=us  
1       https://www.google.com/maps/place//data=!4m2!3m1!1s0x80c2c98c0e3c16fd:0x29ec8a728764fdf9?authuser=-1&hl=en&gl=us  
2       https://www.google.com/maps/place//data=!4m2!3m1!1s0x80c2c778e3b73d33:0xbdc58662a4a97d49?authuser=-1&hl=en&gl=us  
3        https://www.google.com/maps/place//data=!4m2!3m1!1s0x80c2c89923b27a41:0x32041559418d447?authuser=-1&hl=en&gl=us  
4       https://www.google.com/maps/place//data=!4m2!3m1!1s0x80c2c632f933b073:0xc31785961fe826a6?authuser=-1&hl=en&gl=us  
...                                                                                                                  ...  
274996  https://www.google.com/maps/place//data=!4m2!3m1!1s0x88fe64e7daceaaab:0x101c046dcadcc9c3?authuser=-1&hl=en&gl=us  
274997  https://www.google.com/maps/place//data=!4m2!3m1!1s0x88fe59f4602c95fd:0x57e2b1904d9e6949?authuser=-1&hl=en&gl=us  
274998  https://www.google.com/maps/place//data=!4m2!3m1!1s0x88fe71f9617c15e1:0x89518c1d251706bf?authuser=-1&hl=en&gl=us  
274999  https://www.google.com/maps/place//data=!4m2!3m1!1s0x4d34c7517d6e6e29:0x7fe553ee2f090cf0?authuser=-1&hl=en&gl=us  
275000  https://www.google.com/maps/place//data=!4m2!3m1!1s0x88208f4360c053ab:0xe3d6883f19601a01?authuser=-1&hl=en&gl=us  

[275001 rows x 5 columns]

In [ ]:
# Crear un nuevo DataFrame usando loc
dfYeplp_copia = dfYeplp.copy()

Eliminar columnas duplicadas, desde la 14 a 27

In [ ]:
# Crear copia donde solo se excluyen las duplicadas.
dfYeplp_copia = dfYeplp_copia.iloc[:,0:14]

In [ ]:
dfYeplp_copia.info()

<class 'pandas.core.frame.DataFrame'>
Index: 150346 entries, 0 to 150345
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   business_id   150346 non-null  object
 1   name          150346 non-null  object
 2   address       150346 non-null  object
 3   city          150346 non-null  object
 4   state         150343 non-null  object
 5   postal_code   150346 non-null  object
 6   latitude      150346 non-null  object
 7   longitude     150346 non-null  object
 8   stars         150346 non-null  object
 9   review_count  150346 non-null  object
 10  is_open       150346 non-null  object
 11  attributes    136602 non-null  object
 12  categories    150243 non-null  object
 13  hours         127123 non-null  object
dtypes: object(14)
memory usage: 17.2+ MB


Ranking de ciudades con mas comentarios segun Yelp

In [ ]:
rankingFrecuenciaCiudad(dfYeplp_copia, 5)

NameError: name 'rankingFrecuenciaCiudad' is not defined

In [ ]:
# Eliminar las columnas 'url', 'relative_results', 'state' y 'price'
df_cleaned = df1.drop(columns=['url', 'relative_results', 'state', 'price'])

# Ver las primeras filas del DataFrame limpio
print(df_cleaned.head())

               name                                            address  \
0   Porter Pharmacy  Porter Pharmacy, 129 N Second St, Cochran, GA ...   
1      City Textile  City Textile, 3001 E Pico Blvd, Los Angeles, C...   
2      San Soo Dang  San Soo Dang, 761 S Vermont Ave, Los Angeles, ...   
3      Nova Fabrics  Nova Fabrics, 2200 E 11th St, Los Angeles, CA ...   
4  Nobel Textile Co  Nobel Textile Co, 719 E 9th St, Los Angeles, C...   

                                 gmap_id description   latitude   longitude  \
0  0x88f16e41928ff687:0x883dad4fd048e8f8        None  32.388300  -83.357100   
1  0x80c2c98c0e3c16fd:0x29ec8a728764fdf9        None  34.018891 -118.215290   
2  0x80c2c778e3b73d33:0xbdc58662a4a97d49        None  34.058092 -118.292130   
3   0x80c2c89923b27a41:0x32041559418d447        None  34.023669 -118.232930   
4  0x80c2c632f933b073:0xc31785961fe826a6        None  34.036694 -118.249421   

              category  avg_rating  num_of_reviews  \
0           [Pharmacy]    

In [ ]:
print(df_cleaned['MISC'][0])


{'Service options': ['In-store shopping', 'Same-day delivery'], 'Health & safety': ['Mask required', 'Staff required to disinfect surfaces between visits'], 'Accessibility': ['Wheelchair accessible entrance'], 'Planning': ['Quick visit']}


In [ ]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 275001 entries, 0 to 275000
Data columns (total 11 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   name            274994 non-null  object 
 1   address         264939 non-null  object 
 2   gmap_id         275001 non-null  object 
 3   description     13155 non-null   object 
 4   latitude        275001 non-null  float64
 5   longitude       275001 non-null  float64
 6   category        272740 non-null  object 
 7   avg_rating      275001 non-null  float64
 8   num_of_reviews  275001 non-null  int64  
 9   hours           192448 non-null  object 
 10  MISC            194972 non-null  object 
dtypes: float64(3), int64(1), object(7)
memory usage: 23.1+ MB


--------------------------------------------------------------------------------

--------------------------------------------------------------------------------

#  **REVIEWS** YELP

copy paste de la funcion de arriba con la unica diferencia que se separo la funcion de contar los valores duplicados para mejor legibilidad, ya que en este caso se van a estar procesando varios archivos de review

In [27]:
def nulos_vacios_yelp_reviews(df):
    """
    Este procedimiento recibe un DataFrame y muestra la cantidad de valores nulos, vacíos y la cantidad de
    valores duplicados en todas las columnas excepto en las 6, 10, 11 y 13 del DataFrame. También muestra
    el mínimo, máximo y media para columnas de tipo float e integer.

    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    """
    print("Informe de Nulos, Vacíos y Duplicados:")
    print("-" * 30)

    # Total de registros en el DataFrame
    total_registros = len(df)

    # Imprimir total de registros
    print(f"Total de registros: {total_registros}")
    print("-" * 30)

    # Iterar sobre todas las columnas para mostrar nulos y vacíos
    for columna in df.columns:
        # Contar valores nulos
        nulos = df[columna].isnull().sum()

        # Contar valores vacíos (considerando que solo se cuentan si son cadenas)
        vacios = (df[columna] == '').sum() if df[columna].dtype == 'object' else 0

        # Calcular porcentajes
        porcentaje_nulos = (nulos / total_registros) * 100 if total_registros > 0 else 0
        porcentaje_vacios = (vacios / total_registros) * 100 if total_registros > 0 else 0

        # Imprimir resultados
        print(f"Columna: {columna}")
        print(f"  Valores Nulos: {nulos}")
        print(f"  Valores Vacíos: {vacios}")
        print(f"  Porcentaje Nulos: {porcentaje_nulos:.2f}%")
        print(f"  Porcentaje Vacíos: {porcentaje_vacios:.2f}%")

        # Contar y mostrar la cantidad de valores duplicados solo si no es una de las columnas con datos anidados
        # Mostrar mínimo, máximo y media si la columna es de tipo float o int
        if df[columna].dtype in ['float64', 'int64']:
            minimo = df[columna].min()
            maximo = df[columna].max()
            media = df[columna].mean()
            print(f"  Mínimo: {minimo}")
            print(f"  Máximo: {maximo}")
            print(f"  Media: {media:.2f}")
        print("-" * 30)

def duplicados_yelp_reviews(df):
    # Inicializar un diccionario para almacenar los resultados
    resultados = {}

    for columna in df.columns:
        # Revisar si la columna es una de las que queremos analizar
        if columna in [df.columns[7], df.columns[0]]:
            # Filtrar los valores no nulos antes de contar duplicados
            cantidad_duplicados = df[columna].dropna().duplicated(keep=False).sum()
            resultados[columna] = cantidad_duplicados
            print(f"Cantidad de valores duplicados en {columna} (solo no nulos): {cantidad_duplicados}")

    # Devolver el diccionario con los resultados
    return resultados

 **S**e realiza la revision de los archivos de review0,review1, [...], review6

In [ ]:
archivos = 0

while archivos < 7:#dado que son 7 archivos de review
    """
    aca vamos a recorrer archivo por archivo desde el review0 hasta el review6
    en la busqueda de datos vacios o nulos
    """
    #
    ruta = f'/content/drive/MyDrive/Proyecto_final/data/review{archivos}.parquet'
    print("________"*60)
    print("\n"*2)
    print(f"< archivo :  review{archivos}.parquet >")
    df = pd.read_parquet(ruta)
    nulos_vacios_yelp_reviews(df)
    archivos += 1

________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________
< archivo :  review0.parquet >
Informe de Nulos, Vacíos y Duplicados:
------------------------------
Total de registros: 1000000
------------------------------
Columna: review_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
------------------------------
Columna: user_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
------------------------------
Columna: business_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulo

**Como se puede ver en la data, no hay valores nulos o faltantes en cada uno de los archivos revisados de reviews, todas las columnas dan 0 valores faltantes o strings vacios**

Lo que se puede revisar ahora es:
- si hay datos duplicados, mas precisamente si hay review_id duplicados o textos duplicados, ya que seria raro encontrar dos comentarios identicos de dos personas distintas

In [28]:
archivos = 0

while archivos < 7:#dado que son 7 archivos de review
    """
    aca vamos a recorrer archivo por archivo desde el review0 hasta el review6
    en la busqueda de datos duplicados
    """
    #aca vamos a recorrer archivo por archivo desde el review0 hasta el review6
    ruta = f'/content/drive/MyDrive/Proyecto_final/data/review{archivos}.parquet'
    print("________"*60)
    print("\n")
    print(f"< archivo :  review{archivos}.parquet >")
    df = pd.read_parquet(ruta)
    duplicados_yelp_reviews(df)
    archivos += 1

________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________


< archivo :  review0.parquet >
Cantidad de valores duplicados en review_id (solo no nulos): 0
Cantidad de valores duplicados en text (solo no nulos): 3247
__________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________

chequeamos si los duplicados no son mensajes cortos, como 'good service', 'i don't like', etc. o algun otro mensaje generico

In [ ]:
df['text'].value_counts()

text
Still waiting for a call back from the manager or owner......\nSpoke with manager, she had the stylist call me.....\nIf I wanted to deal with her again I would have\nIt's been weeks now.\nTerrible haircut, the worst quality tape in extensions on top of the fact that they were not washed to get the lament coating off before stylist dyed them. Did I mention she got me blonde to match my almost black hair. When questioned, she stated because I had purple added to my hair color the day before. Then why not add purple to dark brown? No answer, color never stayed. \nCompletely washed out, let stylist attempt to re do hair several times and just got worse. Installation is poor at best even though she claimed to be trained in thin hair and tape in extensions.....\nTook twice the time in the morning to try to hide the tape on extension because they were so close to hair line and the crown of my head.\nI wear them because I'm balding from a stressful event in my life......\nIn the end I've gone to drop them off to the stylist 3 times and she is alway gone getting food or not in.\nHer solution is to get them replaced by the company. I've already had to buy new hair thank you. And how can you get them replaced when it your screw up? Plus I don't want to let you touch my head again....\nSo small claims court here we come....\nPrice of hair, all the cuts you charged me for and the useless products you sold me to keep the color in the mess you created.\nI think 3+ months trying to resolve this is more than fair.\nLesson, don't offer services you don't have properly trained to preform...                                                                                                                                                                       4
They moved us 5 years ago about 4 miles. We were  ssatisfied except for the insert for our coffee table that they broke but didn't tell us. We found the parts in the driveway next day. This time we moved from the 2nd floor condo to the 4th floor at a cost of over $700.00 dollars. They were good guys that worked hard. They were about done when one of them came to me and said we had a problem. He said , his words!! I put a small hole in your water bed. I told him i don't even know how to answer that. He showed me the matress  laying on the floor. Water all over the floor. The matress that we slept on the night before was still sitting on the bed after being drained. No water on the bed. They loaded it on a furniture dolly after putting tape on the hole. I told them I could not patch it due to having 3 floors under me. I ask them to take it with them but was told that was not allowed. This is the same bed moved 5 years earlier.we called Mike and was told their was nothing he could do. He would get in touch with his boss Ryan. My wife spoke to Ryan today and was told he spoke to his men and believed them. They say now the hole was already there. How did we sleep on a water matress without a leak the night before. So he is believing them over us. To me that's the same as calling us lires. We were not asking for a new matress but he should have offered something. We even said just give us back the 50 dollar tip. NO. My wife is a christian and would not lie if you set her on fire. I will be talking to a lawyer. I would not deal with a company with low standard's like this 2 men and a truck. They didn't even need the truck. Ryan is a lier and not to be trusted. Stay away from them. I will write again after I speak with my attorney. Terry B    4
One of my favorite places to eat.  They have the best nachos in town. Their fruit tea is pretty good too.                                                                                                                                                                                                                                                                                                                                                                    

##  

---
##  
 - *Dado que son mensajes elaborados lo que vamos a hacer es realizar una comparacion de estos datos para verificar que la fila sea identica*

In [3]:
archivos = 0
lista_duplicado = [] #almacenamos aca los indices que nos va a devolver la funcion duplicated().index.tolist()
while archivos < 7:#dado que son 7 archivos de review
    ruta = f'/content/drive/MyDrive/Proyecto_final/data/review{archivos}.parquet'
    print(f"< archivo :  review{archivos}.parquet >") #simplemente para verificar de que archivo va cargando el proceso
    df = pd.read_parquet(ruta)
    indices_duplicados = df[df['text'].duplicated(keep=False)].index.tolist() # guardamos los indices duplicados
    lista_duplicado.append(indices_duplicados) # los agregamos a la lista
    archivos += 1

< archivo :  review0.parquet >
< archivo :  review1.parquet >
< archivo :  review2.parquet >
< archivo :  review3.parquet >
< archivo :  review4.parquet >
< archivo :  review5.parquet >
< archivo :  review6.parquet >


##  
solo para chequear que hayan valores....

In [9]:
lista_duplicado[0][0:5]

[749, 1283, 2244, 2855, 3124]

se revisa algunas filas para corroborar si los datos de las otras columnas coinciden en su mayoria

In [ ]:
review0 = pd.read_parquet('/content/drive/MyDrive/Proyecto_final/data/review0.parquet')
duplicados = review0.iloc[lista_duplicado[0]]

Vamos a elegir algunos datos al azar

In [ ]:
duplicados

review_id                 user_id  \
749     yCNf7tWAahLZVULW9dEhMw  BdWc0IO_lEjV7FlKv_4oZg   
1283    Aj0UVTXgRLXAasN1X4qXuA  jcgl3ygD81CP-runH1UYRg   
2244    RuPu2dWG2WZIzEV7nSx-qg  MrqeVVEux5nIO3Q76fZFHg   
2855    xgeOvTb78ZxUqrWH1limeA  uSji70QpgHIydaUzigMCgQ   
3124    4F7-40UML8B32d7ApJ76gw  Nwq0AyciH1n9E9Z3AyasLg   
...                        ...                     ...   
993853  SxrEHXbW5FKZCxwJzBAtKg  8uEChvHTHK395lShY2rY3g   
997789  ZR0oJAEW3OAT209ZrypdcQ  DUM2REPlxY8oDHN14x0yow   
997960  TFq85YwKoe-k-Ai8KEgegw  SyABteXA7u_-GA2nKLJ-XQ   
998451  OyG-kVjIHP858iB1OQNNlQ  obMfEZH4L3fjWJBqe2VArg   
998847  Ww8qU3cwb0DFqLFNzIOMWg  FS3vkaoquDXSDoUtT_oJJQ   

                   business_id  stars  useful  funny  cool  \
749     6N9T23_iXkcOCh6boAXpIg      4       0      0     0   
1283    gKPWC7F219oEbu5NXPzH4g      5       0      0     0   
2244    syuKX2pH3JJhOOppzNI6SA      5       1      0     0   
2855    AWnzFqIrlkLAKTc46_zmEg      5       0      0     0   
3124    J5-Y4ujpElrzPGJLYJ4tEg      3       0      0     0   
...                        ...    ...     ...    ...   ...   
993853  qMYCEl7r1K6WYmJglKaryA      4       0      0     0   
997789  TSKCK51GymAsxiA-kx-RiA      2       0      0     0   
997960  Nq1Kn7ud84upBHz9chFW0A      4       1      0     0   
998451  EF57en2jZd9e0m3E-TKV6Q      5       5      1     2   
998847  452karCazKBD2qXw6pLU-Q      5       2      0     0   

                                                     text                date  
749     I've always had good experiences here. The foo... 2015-08-30 19:50:43  
1283    Great food... Love the Fish tacos & Nachos are... 2016-05-07 10:14:52  
2244    We were back last night for the second time: a... 2018-08-04 23:52:52  
2855    I ate like a Queen!  A breakfeast Queen- and I... 2015-06-21 00:05:04  
3124    This is the best pizza by far in Reno. Crisp c... 2011-09-26 16:36:22  
...                                                   ...                 ...  
993853  Came out today to celebrate a friends birthday... 2017-10-20 21:25:50  
997789  Great atmosphere....a beautiful space. Ate din... 2015-05-24 16:29:05  
997960  Yummy sandwiches, live dj on Friday night. Gre... 2016-10-21 19:13:12  
998451  Exceeded my expectations for the price. Staff ... 2015-06-29 20:02:27  
998847  I did a long distance move from Tampa to NJ, t... 2016-07-02 12:26:44  

[3247 rows x 9 columns]

In [ ]:
duplicado = review0[review0['text'].str.contains("Great food... Love the Fish tacos & Nachos")]
duplicado

review_id                 user_id  \
1283    Aj0UVTXgRLXAasN1X4qXuA  jcgl3ygD81CP-runH1UYRg   
106898  5QxU40024gw-5lp9LZLTnQ  jcgl3ygD81CP-runH1UYRg   

                   business_id  stars  useful  funny  cool  \
1283    gKPWC7F219oEbu5NXPzH4g      5       0      0     0   
106898  gKPWC7F219oEbu5NXPzH4g      5       0      0     0   

                                                     text                date  
1283    Great food... Love the Fish tacos & Nachos are... 2016-05-07 10:14:52  
106898  Great food... Love the Fish tacos & Nachos are... 2016-03-21 03:52:16

In [ ]:
duplicado = review0[review0['text'].str.contains("Yummy sandwiches, live dj on Friday night")]
duplicado

review_id                 user_id  \
826111  n4oyGTnUWL7W_k3IHDZ0OA  SyABteXA7u_-GA2nKLJ-XQ   
997960  TFq85YwKoe-k-Ai8KEgegw  SyABteXA7u_-GA2nKLJ-XQ   
999063  620nZUdSFKK_yNLY8cM06g  SyABteXA7u_-GA2nKLJ-XQ   

                   business_id  stars  useful  funny  cool  \
826111  Nq1Kn7ud84upBHz9chFW0A      4       1      0     0   
997960  Nq1Kn7ud84upBHz9chFW0A      4       1      0     0   
999063  Nq1Kn7ud84upBHz9chFW0A      4       0      0     0   

                                                     text                date  
826111  Yummy sandwiches, live dj on Friday night. Gre... 2016-09-11 06:23:47  
997960  Yummy sandwiches, live dj on Friday night. Gre... 2016-10-21 19:13:12  
999063  Yummy sandwiches, live dj on Friday night. Gre... 2016-05-28 05:03:38

**Claramente los datos son casi identicos, usualmente lo unico que suele cambiar es la fecha y el review_id, por lo que podriamos decir que son datos que se ingresaron mal por algun bug de la aplicacion. o por una mala ejecucion durante las consultas de los datos**

#  

#  

#  **USERS** YELP

In [8]:
def duplicados_yelp_users(df):
  """
  esta aplicacion va a recorrer las columnas seleccionadas para verificar si hay datos duplicados
  """
  for columna in df.columns:
    # Verificar si la columna es de tipo objeto (cadena)
    if df[columna].dtype == 'object':
      #la idea es contar si hay review_id o textos duplicados en el df(cosa que no deberia haber)
      if columna in df.columns[0]: # solo vamos a observar la columna 1 de user_id, el resto son caracteristicas que podrian coincidir en varias ocasiones
        # Filtrar los valores no nulos antes de contar duplicados
        cantidad_duplicados = df[columna].dropna().duplicated(keep=False).sum()
        print(f"  Cantidad de valores duplicados en {columna} (solo no nulos): {cantidad_duplicados}")
        return cantidad_duplicados

In [4]:
archivos = 0

while archivos < 3:#dado que son 3 archivos lo seteamos para que recorra los 3 archivos de userX.parquet
    ruta = f'/content/drive/MyDrive/Proyecto_final/data/review{archivos}.parquet'
    print("________"*60)
    print("\n"*2)
    print(f"< archivo :  user{archivos}.parquet >")
    df = pd.read_parquet(ruta)
    nulos_vacios_yelp_reviews(df)
    archivos += 1

________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________



< archivo :  user0.parquet >
Informe de Nulos, Vacíos y Duplicados:
------------------------------
Total de registros: 1000000
------------------------------
Columna: review_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
------------------------------
Columna: user_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
------------------------------
Columna: business_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nul

**En este caso solo vamos a revisar la columna user_id, que no deberia valores duplicados**

In [9]:
archivos = 0

while archivos < 3:#dado que son 3 archivos lo seteamos para que recorra los 3 archivos de userX.parquet
    ruta = f'/content/drive/MyDrive/Proyecto_final/data/review{archivos}.parquet'
    print("________"*60)
    print("\n"*2)
    print(f"< archivo :  user{archivos}.parquet >")
    df = pd.read_parquet(ruta)
    duplicados_yelp_users(df)
    archivos += 1

________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________



< archivo :  user0.parquet >
  Cantidad de valores duplicados en review_id (solo no nulos): 0
______________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________